### ISSUE TITLE
FID misalignment

### ISSUE THREAD

**Issue by ckczzj** [FoundationVision/VAR#45]
Great works and thanks for publishing the code!

I encountered some problems when calculating the FID.

Originally I use my own FID calculation code to calculate the FID between the images from ImageNet validation dataset and their VQVAE autoencoding reconstruction, and the result is 0.92, which is make sense. However, when I use the same code to calculate the FID between the images from ImageNet validation dataset and your **d16** conditional generated images, the result is 19.13. 

I also use your method to calculate the FID: we use the following code to create a npz file and run `python evaluator.py VIRTUAL_imagenet256_labeled.npz tmp.npz`. The result is 18.25. Do your have any idea where I make mistakes.

```
for batch_id, batch in enumerate(imagenet_val_dataloader):
    label = batch["label"]
    gen_images = var.autoregressive_infer_cfg(
                B=label.shape[0],
                label_B=label.to(device),
                cfg=1.5,
                top_k=900,
                top_p=0.96,
                more_smooth=False
            )
    gen_images = gen_images.mul(255).add(0.5).clamp(0, 255).permute(0, 2, 3, 1).to('cpu', torch.uint8).numpy()
    for i in range(gen_images.shape[0]):
        Image.fromarray(gen_images[i]).save(os.path.join("./tmp/", str(batch_id) + "_" + str(i) +  ".png"))

create_npz_from_sample_folder("./tmp")
```

**Comment by keyu-tian** [comment]
@ckczzj could you try it again by re-downloading the model weights from HF and with the latest code? And also load model state dict with var.load_state_dict(var_ckpt, **strict=True**).

**Comment by ckczzj** [comment]
Thanks for replying. Today I re-clone the repo and re-download the ckpt (and load model with `strict=True`) to calculate the FID for **d16** model, and the result is still 18.25.

**Comment by keyu-tian** [comment]
@ckczzj could you save the image into 1000 folders according to their labels (0 to 999)? and then visualize the class 980 (it'll have 50 images) to have a quick check.

**Comment by ckczzj** [comment]
I have solved the problem. Thanks for your replying.

**Comment by Pride-Huang** [comment]
> I have solved the problem. Thanks for your replying.
can you tell me how to solve this problem. FID of var16 is 10.37.

**Comment by Jot-De** [comment]
@ckczzj could you please tell me what was the issue?

**Comment by ckczzj** [comment]
> [@ckczzj](https://github.com/ckczzj) could you please tell me what was the issue?

I used VAR to generate images using DDP, I made a mismake to use the same seed to generate images in different ranks, and the images generated in different ranks are the same. When I set a different seed for different rank, the FID becomes right.

**Comment by Pride-Huang** [comment]
fid needs many different images to compute correctly. same seed makes the same images, not different images.



---Original---
From: "Zijian ***@***.***&gt;
Date: Sat, Feb 22, 2025 10:02 AM
To: ***@***.***&gt;;
Cc: ***@***.******@***.***&gt;;
Subject: Re: [FoundationVision/VAR] FID misalignment (Issue #45)


    
@ckczzj could you please tell me what was the issue?
  
I used VAR to generate images using DDP, I made a mismake to use the same seed to generate images in different ranks, and the images generated in different ranks are the same. When I set a different seed for different rank, the FID becomes right.

—
Reply to this email directly, view it on GitHub, or unsubscribe.
You are receiving this because you commented.Message ID: ***@***.***&gt;
  ckczzj left a comment (FoundationVision/VAR#45)
  
@ckczzj could you please tell me what was the issue?
  
I used VAR to generate images using DDP, I made a mismake to use the same seed to generate images in different ranks, and the images generated in different ranks are the same. When I set a different seed for different rank, the FID becomes right.

**Comment by Jiawei804** [comment]
There are 50k images in the validation set of ImageNet-1k, which are used for both validation during model training and testing for `npz.file`.  Is the FID result reasonable?
